### PR TITLE
fix issue 1074 and clean up junitxml

### LINKED ISSUE #641: MemoryError when writing junitxml for large test suite
Originally reported by: **Michael Howitz (BitBucket: [icemac](http://bitbucket.org/icemac), GitHub: [icemac](http://github.com/icemac))**

----------------------------------------

In a large project (> 3500 Tests) we get a MemoryError when writing the junit xml file.
(The host is our Jenkins CI server with 4 GB of RAM and 2 GB of swap.)

```
#!python

bin/py.test --nosugar --junitxml=parts/tests.xml --cov-report html --cov-report xml --cov-report term --cov src
============================= test session starts ==============================
platform linux2 -- Python 2.7.6 -- py-1.4.25 -- pytest-2.6.4
plugins: gocept.pytestlayer, cache, cov, sugar, remove-stale-bytecode
collected 3643 items

… Many, many tests which ran fine …

Traceback (most recent call last):
  File "bin/py.test", line 308, in <module>
    sys.exit(pytest.main())
  File "/srv/jenkins/eggs/pytest-2.6.4-py2.7.egg/_pytest/config.py", line 41, in main
    return config.hook.pytest_cmdline_main(config=config)
  File "/srv/jenkins/eggs/pytest-2.6.4-py2.7.egg/_pytest/core.py", line 413, in __call__
    return self._docall(methods, kwargs)
  File "/srv/jenkins/eggs/pytest-2.6.4-py2.7.egg/_pytest/core.py", line 424, in _docall
    res = mc.execute()
  File "/srv/jenkins/eggs/pytest-2.6.4-py2.7.egg/_pytest/core.py", line 315, in execute
    res = method(**kwargs)
  File "/srv/jenkins/eggs/pytest-2.6.4-py2.7.egg/_pytest/main.py", line 116, in pytest_cmdline_main
    return wrap_session(config, _main)
  File "/srv/jenkins/eggs/pytest-2.6.4-py2.7.egg/_pytest/main.py", line 109, in wrap_session
    exitstatus=session.exitstatus)
  File "/srv/jenkins/eggs/pytest-2.6.4-py2.7.egg/_pytest/core.py", line 413, in __call__
    return self._docall(methods, kwargs)
  File "/srv/jenkins/eggs/pytest-2.6.4-py2.7.egg/_pytest/core.py", line 424, in _docall
    res = mc.execute()
  File "/srv/jenkins/eggs/pytest-2.6.4-py2.7.egg/_pytest/core.py", line 315, in execute
    res = method(**kwargs)
  File "/srv/jenkins/eggs/pytest-2.6.4-py2.7.egg/_pytest/terminal.py", line 349, in pytest_sessionfinish
    __multicall__.execute()
  File "/srv/jenkins/eggs/pytest-2.6.4-py2.7.egg/_pytest/core.py", line 315, in execute
    res = method(**kwargs)
  File "/srv/jenkins/eggs/pytest-2.6.4-py2.7.egg/_pytest/junitxml.py", line 215, in pytest_sessionfinish
    ).unicode(indent=0))
  File "/srv/jenkins/eggs/py-1.4.25-py2.7.egg/py/_xmlgen.py", line 54, in unicode
    return u("").join(l)
MemoryError
```

Any suggestions?

----------------------------------------
- Bitbucket: https://bitbucket.org/pytest-dev/pytest/issue/641

### LINKED ISSUE #1074: Discard in memory XML representation after done with a node (junitxml)
As commented in #1071, see if it is possible to serialize and discard the xml representation of a node after `teardown` report is processed, in order to conserve memory.

### PR DESCRIPTION
**Pull request by RonnyPfannschmidt**
i started by running yapf on junitxml to get my ide warning-free
did the same on its tests and will simplify the test api a bit (i want the xml api out of the picture)

Fix #1074
Fix #641

### CODE REVIEW

**Comment by RonnyPfannschmidt** [comment]
@alex can you perhaps take a look if that branch has positive effect on the memory usage of the junit reporting of the large testsuite?

**Comment by nicoddemus** [comment]
Sorry for taking so long to review this @RonnyPfannschmidt! :sweat_smile: 

Overall I like the changes you made. As a minor nitpick, I found some PEP-8 errors (mostly too much white-space between functions), but as I said those are minor.

It seems this PR would interest @icemac as well.

File: _pytest/junitxml.py [hunk 1]
```diff
@@ -52,122 +54,94 @@ def repl(matchobj):
...
+        self.properties[name] = bin_xml_escape(value)
 
-    def add_custom_property(self, name, value):
-        self.custom_properties[str(name)] = bin_xml_escape(str(value))
 
-    def _opentestcase(self, report):
-        names = mangle_testnames(report.nodeid.split("::"))
+    def make_properties_node(self):
+        """Return a Junit node containing custom properties set for
+        the current test, if any, and reset the current custom properties.
```
Comment (nicoddemus) [inline _pytest/junitxml.py]: This docstring no longer reflects reality (it doesn't clean `properties`).

File: _pytest/junitxml.py [hunk 2]
```diff
@@ -210,11 +180,107 @@ def append_skipped(self, report):
...
+        self.stats = dict.fromkeys([
+            'error',
+            'passed',
+            'failure',
+            'skipped',
+        ], 0)
+        self.nodereporters = {}  # nodeid -> _NodeReporter
+        self.nodereporters_ordered = []
+
+    def nodereporter(self, nodeid):
```
Comment (nicoddemus) [inline _pytest/junitxml.py]: `node_reporter` perhaps?

File: _pytest/junitxml.py [hunk 3]
```diff
@@ -210,11 +180,107 @@ def append_skipped(self, report):
...
+        logfile = os.path.expanduser(os.path.expandvars(logfile))
+        self.logfile = os.path.normpath(os.path.abspath(logfile))
+        self.prefix = prefix
+        self.stats = dict.fromkeys([
+            'error',
+            'passed',
+            'failure',
+            'skipped',
+        ], 0)
+        self.nodereporters = {}  # nodeid -> _NodeReporter
```
Comment (nicoddemus) [inline _pytest/junitxml.py]: `node_reporters` and `node_reporters_ordered`?

File: _pytest/junitxml.py [hunk 4]
```diff
@@ -210,11 +180,107 @@ def append_skipped(self, report):
...
+    The fixture is callable with (name, value), with value being automatically
+    xml-encoded.
+    """
+    request.node.warn(
+        code='C3',
+        message='record_xml_property is an experimental feature',
+    )
+    xml = getattr(request.config, "_xml", None)
+    if xml is not None:
+        nodereporter = xml.nodereporter(request.node.nodeid)
```
Comment (nicoddemus) [inline _pytest/junitxml.py]: Well done, cleaner than the previous impl

File: _pytest/junitxml.py [hunk 5]
```diff
@@ -210,11 +180,107 @@ def append_skipped(self, report):
...
+        reporter = _NodeReporter(nodeid, self)
+        self.nodereporters[nodeid] = reporter
+        self.nodereporters_ordered.append(reporter)
+        return reporter
+
+    def add_stats(self, key):
+        if key in self.stats:
+            self.stats[key] += 1
+
+    def _opentestcase(self, report):
```
Comment (nicoddemus) [inline _pytest/junitxml.py]: perhaps this should be named `_record_report` instead?

**Comment by nicoddemus** [comment]
Aside from my minor comments, LGTM. :+1: Feel free to merge if you think my naming suggestions are not worthwhile to make right now. :smile:

File: _pytest/junitxml.py (hunk 3 above)
Comment (RonnyPfannschmidt) [inline _pytest/junitxml.py]: Lookup vs serialisation order

Comment (RonnyPfannschmidt) [inline _pytest/junitxml.py]: Basically it can be replaced with a ordereddict

Comment (nicoddemus) [inline _pytest/junitxml.py]: Sorry I was not clear, I just mentioning that perhaps they should be renamed to `node_reporters` and `node_reporters_ordered`. My bad. :stuck_out_tongue_closed_eyes:

**Comment by icemac** [comment]
I will have a look after this PR is merged and released. Thanks for looking into it.

**Comment by RonnyPfannschmidt** [comment]
@icemac please take a look at it beforehand, i don't have a good benchmark atm, and i'd like to be sure i'm delivering the right thing

**Comment by icemac** [comment]
@RonnyPfannschmidt I'll test today.

**Comment by icemac** [comment]
@RonnyPfannschmidt I get an exception after some tests are run (summing up the number of successes etc. in the summary line is less than the number of collected items), see the transcript:
```
bin/py.test -p no:sugar --junitxml=parts/tests.xml --cov-report html --cov-report xml --cov-report term --cov src
============================= test session starts ==============================
platform linux2 -- Python 2.7.9, pytest-2.8.2.dev1, py-1.4.30, pluggy-0.3.1
rootdir: .../instance, inifile: pytest.ini
plugins: gocept.pytestlayer-2.0, cov-1.8.1, remove-stale-bytecode-1.0
collected 4449 items
[… many tests run here …]
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File ".../src/pytest/_pytest/main.py", line 90, in wrap_session
INTERNALERROR>     session.exitstatus = doit(config, session) or 0
INTERNALERROR>   File ".../src/pytest/_pytest/main.py", line 121, in _main
INTERNALERROR>     config.hook.pytest_runtestloop(session=session)
INTERNALERROR>   File ".../src/pytest/_pytest/vendored_packages/pluggy.py", line 724, in __call__
INTERNALERROR>     return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
INTERNALERROR>   File ".../src/pytest/_pytest/vendored_packages/pluggy.py", line 338, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File ".../src/pytest/_pytest/vendored_packages/pluggy.py", line 333, in <lambda>
INTERNALERROR>     _MultiCall(methods, kwargs, hook.spec_opts).execute()
INTERNALERROR>   File ".../src/pytest/_pytest/vendored_packages/pluggy.py", line 596, in execute
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File ".../src/pytest/_pytest/main.py", line 146, in pytest_runtestloop
INTERNALERROR>     item.config.hook.pytest_runtest_protocol(item=item, nextitem=nextitem)
INTERNALERROR>   File ".../src/pytest/_pytest/vendored_packages/pluggy.py", line 724, in __call__
INTERNALERROR>     return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
INTERNALERROR>   File ".../src/pytest/_pytest/vendored_packages/pluggy.py", line 338, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File ".../src/pytest/_pytest/vendored_packages/pluggy.py", line 333, in <lambda>
INTERNALERROR>     _MultiCall(methods, kwargs, hook.spec_opts).execute()
INTERNALERROR>   File ".../src/pytest/_pytest/vendored_packages/pluggy.py", line 595, in execute
INTERNALERROR>     return _wrapped_call(hook_impl.function(*args), self.execute)
INTERNALERROR>   File ".../src/pytest/_pytest/vendored_packages/pluggy.py", line 253, in _wrapped_call
INTERNALERROR>     return call_outcome.get_result()
INTERNALERROR>   File ".../src/pytest/_pytest/vendored_packages/pluggy.py", line 279, in get_result
INTERNALERROR>     _reraise(*ex)  # noqa
INTERNALERROR>   File ".../src/pytest/_pytest/vendored_packages/pluggy.py", line 264, in __init__
INTERNALERROR>     self.result = func()
INTERNALERROR>   File ".../src/pytest/_pytest/vendored_packages/pluggy.py", line 596, in execute
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File ".../src/pytest/_pytest/runner.py", line 65, in pytest_runtest_protocol
INTERNALERROR>     runtestprotocol(item, nextitem=nextitem)
INTERNALERROR>   File ".../src/pytest/_pytest/runner.py", line 72, in runtestprotocol
INTERNALERROR>     rep = call_and_report(item, "setup", log)
INTERNALERROR>   File ".../src/pytest/_pytest/runner.py", line 123, in call_and_report
INTERNALERROR>     hook.pytest_runtest_logreport(report=report)
INTERNALERROR>   File ".../src/pytest/_pytest/vendored_packages/pluggy.py", line 724, in __call__
INTERNALERROR>     return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
INTERNALERROR>   File ".../src/pytest/_pytest/vendored_packages/pluggy.py", line 338, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File ".../src/pytest/_pytest/vendored_packages/pluggy.py", line 333, in <lambda>
INTERNALERROR>     _MultiCall(methods, kwargs, hook.spec_opts).execute()
INTERNALERROR>   File ".../src/pytest/_pytest/vendored_packages/pluggy.py", line 596, in execute
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File ".../src/pytest/_pytest/junitxml.py", line 320, in pytest_runtest_logreport
INTERNALERROR>     self.update_testcase_duration(report)
INTERNALERROR>   File ".../src/pytest/_pytest/junitxml.py", line 329, in update_testcase_duration
INTERNALERROR>     reporter.duration += getattr(report, 'duration', 0.0)
INTERNALERROR> AttributeError: '_NodeReporter' object has no attribute 'duration'
28 failed, 2178 passed, 73 skipped, 10 xfailed, 9 pytest-warnings in 400.96 seconds 
```

**Comment by RonnyPfannschmidt** [comment]
i see, good find, i'll investigate

**Comment by RonnyPfannschmidt** [comment]
after investigation i'm certain the issue cannot be replicated when running the correct version of pytest

@nicoddemus @hpk42 this one seems ready for merge

File: _pytest/junitxml.py [hunk 6]
```diff
@@ -210,11 +179,107 @@ def append_skipped(self, report):
...
+        logfile = os.path.expanduser(os.path.expandvars(logfile))
+        self.logfile = os.path.normpath(os.path.abspath(logfile))
+        self.prefix = prefix
+        self.stats = dict.fromkeys([
+            'error',
+            'passed',
+            'failure',
+            'skipped',
+        ], 0)
+        self.nodere_porters = {}  # nodeid -> _NodeReporter
```
Comment (nicoddemus) [inline _pytest/junitxml.py]: Should this be `node_reporters`?

Comment (RonnyPfannschmidt) [inline _pytest/junitxml.py]: good find

**Comment by nicoddemus** [comment]
Could be that the problem reported by @icemac be due to `_NodeReporter.finalize()` being called twice? That would explain the `AttributeError`.

**Comment by RonnyPfannschmidt** [comment]
i think its related to the issue about duplicate tests

**Comment by nicoddemus** [comment]
Good idea. But then it should be simple to verify it, just pass the same file name to pytest twice and you should see the same error. If that's the case we should fix this branch, as we just decided that running the same set of tests more than once is a valid use case.

**Comment by RonnyPfannschmidt** [comment]
i just tried, and it will err out

**Comment by nicoddemus** [comment]
The current master will repeat a new `testcase` tag when you execute the same test twice:

```xml
<?xml version="1.0" encoding="utf-8"?>
<testsuite errors="0" failures="0" name="pytest" skips="0" tests="2" time="0.013">
  <testcase classname="tmp.test_foo1" file="tmp\test_foo1.py" line="0" name="test_1" time="0.0010001659393310547"/>
  <testcase classname="tmp.test_foo1" file="tmp\test_foo1.py" line="0" name="test_1" time="0.0010001659393310547"/>
</testsuite>
```

I was thinking, to support the same case here, perhaps we could:

* instead of a `node_reporters_ordered`, just keep a list of xml strings which have been finished so far (`_NodeReporter.to_xml()`); lets call that `xml_test_cases`;
* instead of calling `_NodeReporter.finalize()` when you see a `"teardown"` event in `pytest_runtest_logreport`, just call `to_xml()`, store it in `xml_test_cases` and discard the `_NodeReporter` instance for that `nodeid`. We won't even need a `finalize` method anymore.

This will change the order of the `testsuite` tags in the report, where previously they would appear in the order they were "setup" first, but now they will appear in the order they where "teardown" first, but I don't think this is an issue at all.

What do you think?

**Comment by RonnyPfannschmidt** [comment]
Will break on xdist

**Comment by nicoddemus** [comment]
> Will break on xdist

You mean in the normal usage, or for duplicated test ids?

**Comment by RonnyPfannschmidt** [comment]
For duplicate tests, they can interleave wrt reports

**Comment by nicoddemus** [comment]
Thought about that too, but that happens today already IIUC (both `master` and this PR)... my proposal would at least avoid the internal error.

**Comment by RonnyPfannschmidt** [comment]
I plan to fix both by extending the key with the slave ID and implemening your proposal

**Comment by nicoddemus** [comment]
Sounds good, thanks!

But how will you know which item was executed in which slave? We don't have how to get that information now, I think.

**Comment by nicoddemus** [comment]
Oooh didn't know about this line:

```python
    def slave_testreport(self, node, rep):
        if rep.when == "call" or (rep.when == "setup" and not rep.passed):
            self.sched.remove_item(node, rep.item_index, rep.duration)
        # self.report_line("testreport %s: %s" %(rep.id, rep.status))
        rep.node = node   # <----
        self.config.hook.pytest_runtest_logreport(report=rep)
        self._handlefailures(rep)
```

So the `report` object knows where each test executed, nice. Your suggestion seems perfect then. :cool:

**Comment by RonnyPfannschmidt** [comment]
that little gem came directly from @hpk42  :)

**Comment by RonnyPfannschmidt** [comment]
@nicoddemus can you also take a look at this one again?

**Comment by nicoddemus** [comment]
Oh sorry, I didn't notice you have pushed new commits.

**Comment by nicoddemus** [comment]
I think this change warrants a CHANGELOG entry

**Comment by nicoddemus** [comment]
I think after you add a CHANGELOG entry you can merge this @RonnyPfannschmidt

**Comment by RonnyPfannschmidt** [comment]
im on it

**Comment by RonnyPfannschmidt** [comment]
rebased on to of master and added the entry, i'll wait for travis then push the button

**Comment by nicoddemus** [comment]
:+1:

**Pull request merged**